addons/html_editor/static/src/others/embedded_components/plugins/toggle_plugin/toggle_plugin.js

```diff
@@ -0,0 +1,447 @@
+        newParagraph.textContent = closestToggleTitle.textContent || "";
+        if (!newParagraph.textContent) {
+            newParagraph.appendChild(this.document.createElement("br"));
+        }
+        const contentToInsert = container.querySelectorAll(`${contentSelector} > *`);
+        if (contentToInsert.length > 1 || !isEmptyBlock(contentToInsert[0])) {
+            const [rightNode, rightOffset] = rightPos(container);
+            this.dependencies.selection.setSelection({
+                anchorNode: rightNode,
+                anchorOffset: rightOffset,
+                focusNode: rightNode,
+                focusOffset: rightOffset,
+            });
+            const fragment = this.document.createDocumentFragment();
+            fragment.replaceChildren(...contentToInsert);
+            this.dependencies.dom.insert(fragment);
+        }
+        container.replaceWith(newParagraph);
+        this.dependencies.history.addStep();
+        this.dependencies.selection.setCursorStart(newParagraph);
+        return true;
+    }
+
+    /**
+     * Handles the tab behavior. This means that when we are inside a toggle title and we have a toggle
+     * as previous sibling of the embedded component, the current toggle is indented inside the content of
+     * the previous one.
+     */
+    handleTab() {
+        const container = this.getToggleContainer();
+        if (container) {
+            // If selection is in a title.
+            const previousSibling = container.previousElementSibling;
+            if (previousSibling?.matches("[data-embedded='toggle']")) {
+                // If the previous element of the embedded component is also a toggle, we need to handle
+                // it.
+                const { restore: restoreSelection } =
+                    this.dependencies.selection.preserveSelection();
+                this.dependencies.selection.setCursorEnd(
+                    previousSibling.querySelector("[data-embedded-editable='content']")
+                );
+                previousSibling.firstChild.querySelector("i.fa-caret-right")?.click(); // open the toggle if needed
+                const fragment = this.document.createDocumentFragment();
+                fragment.replaceChildren(container);
+                const canReplace = previousSibling.querySelector(
+                    "[data-embedded-editable='content'] > *:only-child"
+                );
+                // If the 1st block of the previous content is empty we replace it with our toggle.
+                // Else we add it to the content.
+                if (isEmptyBlock(canReplace)) {
+                    canReplace.replaceWith(fragment);
+                } else {
+                    this.dependencies.dom.insert(fragment);
+                }
+                this.dependencies.history.addStep();
+                window.setTimeout(restoreSelection, "animationFrame"); // Used to handle caret displaying issues
```

Comment:
The issues you have are caused by the changed state of the toggle.
Using `animationFrame` is indeterministic.

What I'm suggesting instead:
`useExternalListener` on a custom event (i.e. `forceShowToggle`) on the host in `toggle.js`.
In the plugin, dispatch the event instead of `click`, and provide `restoreSelection` as a callback

in `toggle.js`, the handler for the `forceShowToggle` should change the state if it is not already valid, and we should also have a `useEffect` with `restoreSelection` as a dependency (the second argument of the useEffect)
=> if the state changes, and the `restoreSelection` function has changed since the last time, we call `restoreSelection` if it is not undefined.

Let me know if you have any question.